### PR TITLE
renaming for clarity

### PR DESCRIPTION
**Pull request by mschoch**

### CODE REVIEW

**Comment by mschoch** [comment]
I have some other ideas too, but these are probably already controversial, so I'll start here.

**Comment by mschoch** [comment]
Summary:

- batch things moved to batch.go
- optimize things moved to optimize.go
- prefix Index in names like IndexXYZ removed (Go best practice as the name "stutters" because the pkg name is already index)
- that leaves problem for Index, as there is nothing left.  however I feel upon thinking about it, this object really is used as a writer, hence the name

**Review by abhinavdangeti — APPROVED** [review]

File: index.go [hunk 1]
```diff
@@ -44,15 +43,15 @@ type Index interface {
 
 	// Reader returns a low-level accessor on the index data. Close it to
 	// release associated resources.
-	Reader() (IndexReader, error)
+	Reader() (Reader, error)
```
Comment (sreekanth-cb) [inline index.go:46]: It is taking time for me to follow the part that this - Writer interface having a Reader().. 
reads like the writer implementation has to return a reader instance isn't a very convincing point?

**Review by sreekanth-cb — COMMENTED** [review]
.

File: index.go (hunk 1 above)
Comment (mschoch) [reply index.go:46]: So, here is how I would try to convince you.

First, yes, having a Writer() return a Reader() at first seems awkward, and one might claim that typically Go APIs do not look like this.

However, I would then say, think more about the *kind* of Reader that we return.  This reader can search data that *only exists* in the memory structures of *that* Writer.  We do this, because it is very useful for applications that want to offer lower latency searching (latency in this context meaning index->search availability)  Now, if you think about this capability that these Readers have, they could *only* be implemented by a *Writer*, because the returned Reader must have access to the Writer's internal state.

Next, the reason it seems more weird is that this is the *only* kind of Reader that Bleve offers today.  But, we only have to look to Bluge to see that it's possible to offers other kinds of Readers.  Specially, ones that cannot search the in-memory structures of the Writer, and thus can be opened directly, and not by first acquiring a Writer.

So, in summary, my argument would be that, the weird thing is that Bleve only suppports that one kind of Reader (which requires first getting a Writer).  That a Writer can return a Reader, is not that weird, if you consider it's capabilities.

Comment (sreekanth-cb) [reply index.go:46]: > However, I would then say, think more about the _kind_ of Reader that we return. This reader can search data that _only exists_ in the memory structures of _that_ Writer.

Not sure whether I get this correctly. My thinking goes like - with the copy-on-write/mvcc sort of techniques the writer implementations use today, the reader obtained from writer.Reader() is just a glimpse of the writer's state at that point in time and never gives any assurance for any sort of consistency between their (reader-writer) states. 

If so, then it only confuses the intent of the interface IMHO.
But I can imagine this could work well for a static builder (index once and reader thereafter) interface.

Comment (mschoch) [reply index.go:46]: I don't think I'm arguing any assurance about consistency between the reader and writer states.

I'm thinking about the system in these terms:

Index - an abstract concept, it does not have any direct interface
Writer - the interface to altering the index in any way
Reader - an interface to searching the index

Bleve could have an OpenReader() or whatever you want to cal some top-leve Reader() method that does not come from a Writer, but it simply doesn't offer this (Bluge does).

Instead, it's only offering is a Reader which also searches through any in-memory items inside the Writer.  Because of this, it has to come from the Writer.

At this point I think we're just disagreeing on the terminology.  It seems as you take a different perspective.  The in-flight memory items are not part of the Writer, but instead part of the index.  That was how I used to think about things, but I found it limiting because then the Index just has 2 methods, Reader() and Writer().  But, what is the writer?  It's all of the scorch package, because the entire thing is just one big complex index writing implementation.

Comment (mschoch) [reply index.go:46]: Another point I'd like to make.  Eventually, I would expect that Reader() on the Writer interface would move out and become an optional additional interface.  The reason is that many much-simpler index.Writer implementations would not be able to support it.

The problem is that in Scorch, that is the only type of Reader that we support.

Comment (sreekanth-cb) [reply index.go:46]: >"The in-flight memory items are not part of the Writer, but instead part of the index"

Yes, and the name “Index” was looking better to me there. 
When we replace Index with a Writer interface - those two are not equivalent/interchangable notions in mind.
And when the Reader comes from a Writer, it tends to give a wrong sense of strong write/read consistency/some tighter relation for the beholder. (since fresh eyes don't expect a reader from a writer).

With the original Index as an interface, the reader bond seems a lighter one.


> Another point I'd like to make. Eventually, I would expect that Reader() on the Writer interface would move out and become an optional additional interface. The reason is that many much-simpler index.Writer implementations would not be able to support it.

I would prefer that, as I might have otherwise missed the urgency of the current change. 


Not yet convinced about the renaming of “Index => Writer” as this isn’t brining any clarity.
To follow this new naming convention, one need to be aware of implementation details of it, like
Why a reader from a writer makes sense ?   Is that a sign of better naming, not sure..

Comment (mschoch) [reply index.go:46]: I see, well I tried, but we have to keep moving.  I'll close this.
